### PR TITLE
feat: add flex tier pricing support with flat rate structure

### PR DESCRIPTION
**Pull request by Pratham-Mishra04**
## Summary

This PR adds support for OpenAI's flex service tier pricing alongside the existing priority tier support. The flex tier provides flat-rate pricing that ignores token count thresholds, while priority tier continues to use tiered pricing based on token counts.

## Changes

- Added flex tier pricing columns to the database schema (`input_cost_per_token_flex`, `output_cost_per_token_flex`, `cache_read_input_token_cost_flex`)
- Implemented flex tier pricing logic that applies flat rates regardless of token count
- Refactored service tier handling from a simple boolean to a structured `serviceTier` type supporting both priority and flex tiers
- Added database migration for flex tier pricing columns with proper rollback support
- Updated pricing calculation functions to handle flex tier across all request types (chat, embedding, rerank, speech, transcription, image, video)
- Enhanced test coverage for flex tier scenarios and edge cases

## Type of change

- [x] Feature

## Affected areas

- [x] Core (Go)

## How to test

Validate the pricing calculations work correctly for flex tier:

```sh
# Run the pricing tests
go test ./framework/modelcatalog -v -run TestComputeTextCost_Flex
go test ./framework/modelcatalog -v -run TestCalculateCost_FlexTier

# Run all pricing tests to ensure no regressions
go test ./framework/modelcatalog -v -run TestComputeTextCost
go test ./framework/modelcatalog -v -run TestTieredCacheReadRate

# Test database migrations
go test ./framework/configstore -v -run TestMigration
```

Test with actual OpenAI responses that include `service_tier: "flex"` to verify end-to-end pricing calculation.

## Breaking changes

- [ ] Yes
- [x] No

The changes are backward compatible. Existing priority tier functionality remains unchanged, and flex tier is additive.

## Security considerations

No security implications. This change only affects pricing calculations and database schema for storing pricing data.

## Checklist

- [x] I read `docs/contributing/README.md` and followed the guidelines
- [x] I added/updated tests where appropriate
- [x] I updated documentation where needed
- [x] I verified builds succeed (Go and UI)
- [x] I verified the CI pipeline passes locally if applicable

### CODE REVIEW

**Comment by Pratham-Mishra04** [comment]
* **#2641** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2641?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2641?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2640** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2640?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: framework/modelcatalog/pricing.go [hunk 1]
```diff
@@ -698,56 +726,59 @@ func tieredVideoInputPerSecondRate(pricing *configstoreTables.TableModelPricing,
...
+func tieredCacheCreationInputTokenRate(pricing *configstoreTables.TableModelPricing, totalTokens int, tier serviceTier) float64 {
 	if totalTokens > TokenTierAbove200K && pricing.CacheCreationInputTokenCostAbove200kTokens != nil {
 		return *pricing.CacheCreationInputTokenCostAbove200kTokens
 	}
 	if pricing.CacheCreationInputTokenCost != nil {
 		return *pricing.CacheCreationInputTokenCost
 	}
-	return tieredInputRate(pricing, totalTokens, isPriority)
+	return tieredInputRate(pricing, totalTokens, tier)
 }
```
[the remark this answers is not in the document]
Comment (Pratham-Mishra04) [reply framework/modelcatalog/pricing.go:785]: @greptile this is because flex is only openai compatible pricing and openai has no cache creation construct, only cache read

[the remark this answers is not in the document]
Comment (Pratham-Mishra04) [reply framework/modelcatalog/pricing.go:785]: @greptile added a comment to `tieredCacheCreationInputTokenRate` clarifying the intent — flex isn't checked there because OpenAI (the only flex-tier provider) has no cache creation concept, so the guard would be dead code. The asymmetry is now documented inline.

File: framework/modelcatalog/pricing_test.go [hunk 2]
```diff
@@ -1658,46 +1658,268 @@ func TestTieredCacheReadRate_FallbackOrder(t *testing.T) {
...
+	}
+
+	cost := computeTextCost(&p, usage, serviceTier{isFlex: true})
+
+	// Non-cached input: (1000-400)*0.0000015 = 600*0.0000015 = 0.0009
+	// Cached read (flex rate): 400*0.0000006 = 0.00024
+	// Output: 500*0.0000075 = 0.00375
+	// Total: 0.0009 + 0.00024 + 0.00375 = 0.00489
+	assert.InDelta(t, 0.00489, cost, 1e-12)
+}
```
[the remark this answers is not in the document]
Comment (Pratham-Mishra04) [reply framework/modelcatalog/pricing_test.go:1845]: @greptile skipping this one — the scenario (flex tier + `CachedWriteTokens > 0`) can't occur in practice since OpenAI (the only flex-tier provider) has no cache creation concept, so `CachedWriteTokens` will always be 0 on flex responses. A test for an impossible input would be noise rather than a regression guard.

**Comment by akshaydeo** [comment]
### Merge activity

* **Apr 11, 7:53 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/2641).
* **Apr 11, 7:54 AM UTC**: [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/2641) couldn't merge this PR because it had merge conflicts.
* **Apr 11, 8:13 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/2641).
* **Apr 11, 8:13 AM UTC**: @akshaydeo merged this pull request with [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/2641).
